Add Content_script manifest key (#6981)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/manifest/content_scripts/index.md

---
layout: 'layouts/doc-post.njk'

> title: 'Manifest - Content Scripts'

Reviewer: ```suggestion
title: 'Manifest - content scripts'
```

Reviewer: Here's my reasoning. Titles are sentence case. There is no rule for capitalization after a dash or hyphen, or at leaste none that I can find. However, there is a rule for capitalization after a colon. (Chicago Manual of Style, 17th ed., rule 6.63.) Hyphens and colons have overlapping and analogous functions, so 6.63 applies.

> seoTitle: 'Chrome Extensions Manifest: content_scripts'

Reviewer: ```suggestion
seoTitle: 'Chrome Extensions Manifest: "content_scripts"'
```

date: 2023-08-10

> description: Reference documentation for the content_scripts property of manifest.json.

Reviewer: ```suggestion
description: Reference documentation for the "content_scripts" property of manifest.json.
```

---

## Overview {: #overview }

> The `"content_scripts"` field is used to statically load a JavaScript or CSS file every time a page is opened that matches a certain [URL pattern][doc-match]. Extensions can also inject content scripts programmatically, see [Injecting Scripts][doc-cs] for details.

Reviewer: ```suggestion
The `"content_scripts"` field specifies a statically loaded JavaScript or CSS file to be used every time a page is opened that matches a certain [URL pattern][doc-match]. Extensions can also inject content scripts programmatically, see [Injecting Scripts][doc-cs] for details.
```

Reviewer: Active voice.

> The `"content_scripts"` field is used to statically load a JavaScript or CSS file every time Chrome opens a page that matches a certain [URL pattern][doc-match]. Extensions can also inject content scripts programmatically, see [Injecting Scripts][doc-cs] for details.

Reviewer 2: Nit: I feel like the user opens a page, not Chrome.

```suggestion
The `"content_scripts"` field is used to statically load a JavaScript or CSS file every time a page is opened that matches a certain [URL pattern][doc-match]. Extensions can also inject content scripts programmatically, see [Injecting Scripts][doc-cs] for details.
```

## Manifest {: #manifest }

> These are the supported fields for `"content_scripts"`. Only the `"matches"` key is required.

Author: I may have borrowed this from M...N, but looking at the [source code](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/manifest_handlers/content_scripts_handler.cc;drc=37336012d6551ad82eae1c428c997071253dbf3e;l=158) I see you are correct.... Must follow the source of all truth!!! ✨

{% Label %}manifest.json{% endLabel %}

```json
{
 "name": "My extension",
 ...
 "content_scripts": [
   {
     "matches": ["https://*.example.com/*"],
     "css": ["my-styles.css"],
     "js": ["content-script.js"],
     "exclude_matches": ["*://*/*foo*"],
     "include_globs": ["*example.com/???s/*"],
     "exclude_globs": ["*bar*"],     
     "all_frames": false,
     "match_origin_as_fallback": false,
     "match_about_blank": false,
     "run_at": "document_idle",
     "world": "ISOLATED",
   }
 ],
 ...
}
```

## Files {: #files }

> Each file must contain a relative path to a resource in the extension's root directory. Leading slashes (`/`) are automatically trimmed. The [`"run_at"`](#world-timings) specifies when each file will be injected.  

Reviewer: ```suggestion
Each file must contain a relative path to a resource in the extension's root directory. Leading slashes (`/`) are automatically trimmed. The [`"run_at"`](#world-timings) key specifies when each file will be injected.  
```

`"css"` - Array

> : _Optional_. An array of CSS file paths, injected in the order of this array, and before any DOM construction or page rendering occurs.

Reviewer: >_Optional_.

Is this an existing idiom? The italics are good for making it scannable. I want to try to remember that you did it this way.

      
`"js"` - Array,
: _Optional_. An array of JavaScript file paths, injected in the order they appear in this array, after css files are injected.

## Matching URLs {: #match-urls }

Only the `"matches"` property is required. Then you can use `"exclude_matches"`, `"include_globs"`, and `"exclude_globs"` to customize which URLs to inject code into. The `"matches"` key will [trigger a warning][perm-warn]. 

`"matches"` - Array

> : _Required_. Specifies which URL patterns to inject the content scripts into. See [Match Patterns][doc-match] for syntax.

Reviewer: I know it's actually pages that scripts are getting injected into, not URLs, but I can't think of rewrite that isn't confusing. Am I overthinking this?

`"exclude_matches"` - Array
: _Optional_. Excludes the URL patterns to inject the content scripts into. See [Match Patterns][doc-match] for syntax.

`"include_globs"` - Array

> : _Optional_. Applied after matches to include only those URLs that also match this glob. Intended to emulate the [@include][grease-include] Greasemonkey keyword.

Reviewer: If I understand this correctly:
1. I have a set of matching URLs.
2. Then I can alter the resulting set with include/exclude globs.

Is this correct?

> : _Optional_. Applied after matches to include only those URLs that also match this glob. Intended to emulate the [@include][grease-include] Greasemonkey keyword.

Author: There's few things I'm comfortable with... 🥸

> : _Optional_. Applied after matches to include only those URLs that also match this glob. Intended to emulate the [@include][grease-include] Greasemonkey keyword.

Reviewer 2: Fair 🤷

        
`"exclude_globs"` - Array
: _Optional_. Applied after matches to exclude URLs that match this glob. Intended to emulate the [@exclude][grease-exclude] Greasemonkey keyword.

> Glob string URLs can contain "wildcards" **\*** and question marks. The wildcard **\*** matches any string of any length, including an empty string, while the question mark **?** matches any single character. 

Reviewer: ```suggestion
Glob URLs are those that contain "wildcards" **\*** and question marks. The wildcard **\*** matches any string of any length, including an empty string, while the question mark **?** matches any single character. 
```

Reviewer: This is actually the [definition of 'glob'](https://en.wikipedia.org/wiki/Glob_(programming))

> The content script is injected to a page if:

Reviewer: ```suggestion
The content script is injected into a page if:
```

- Its URL matches any `"matches"` and `"include_globs"` patterns.
- And the URL doesn't match `"exclude_matches"` or `"exclude_globs"` patterns.

{% Details %}
{% DetailsSummary %}
### Globs and URL matching examples {: #matching-samples }
{% endDetailsSummary %}

#### `"include_globs"` {: #incl-globs }

{% Label %}manifest.json{% endLabel %}

> ```json/5

Reviewer: Actually, this means highlight line 5.

  "content_scripts": [
    {
      "matches": ["https://*.nytimes.com/*"],
      "include_globs": ["*nytimes.com/???s/*"],
      "js": ["content-script.js"]
    }
  ],
  ...
}
```

<div class="switcher">
{% Compare 'better', 'Matches' %}
```text
https://www.nytimes.com/arts/index.html
https://www.nytimes.com/jobs/index.html
```
{% endCompare %}

{% Compare "worse", "Does not match" %}
```text
https://www.nytimes.com/sports/index.html
https://www.nytimes.com/music/index.html
```
{% endCompare %}
</div>

#### `"exclude_globs"` {: #excl-globs }

{% Label %}manifest.json{% endLabel %}

```json/5
{
  ...
  "content_scripts": [
    {
      "matches": ["https://*.nytimes.com/*"],
      "exclude_globs": ["*science*"],
      "js": ["content-script.js"]
    }
  ],
  ...
}
```

<div class="switcher">
{% Compare 'better', 'Matches' %}
```text
https://history.nytimes.com
https://.nytimes.com/music
```
{% endCompare %}

{% Compare "worse", "Does not match" %}
```text
https://science.nytimes.com
https://www.nytimes.com/science
```
{% endCompare %}
</div>

#### Advanced customization example {: #all-custom }

{% Label %}manifest.json{% endLabel %}

```json/5-7
{
  ...
  "content_scripts": [
    {

>       "matches": ["https://*.nytimes.com/*"],

Author: I don't have a strong opinion on this. Maybe the original idea was that the nytimes has different topics so the glob examples would match?

If we switch to example.com, would be only use foo and bar? or also use the topics

>       "matches": ["https://*.nytimes.com/*"],

Reviewer 2: I didn't realise we had this already. I'm not personally a huge fan, but don't feel strongly :)

Yeah, if we used example.com I think we could still use topics.

      "exclude_matches": ["*://*/*business*"],
      "include_globs": ["*example.com/???s/*"],
      "exclude_globs": ["*science*"],
      "js": ["content-script.js"]
    }
  ],
  ...
}
```

<div class="switcher">
{% Compare 'better', 'Matches' %}
```text
https://www.example.com/arts/index.html
https://.example.com/jobs/index.html
```
{% endCompare %}

{% Compare "worse", "Does not match" %}
```text
https://science.example.com
https://www.example.com/jobs/business
https://www.example.com/science
```
{% endCompare %}
</div>

{% endDetails %}

## Frames {: #frames }

> The `"all_frames"` field specifies if the content script should be injected into all frames matching the specified URL requirements. If set to `false` it will only inject into the topmost frame. It can be used along with `"match_about_blank"` to inject into an `about:blank` frame. 

Reviewer: How strict should we be about using 'key' instead of 'field'?

> The `"all_frames"` field specifies if the content script should be injected into all frames matching the specified URL requirements. If set to `false` it will only inject into the topmost frame. It can be used along with `"match_about_blank"` to inject into an `about:blank` frame. 

Reviewer: >If set to `false`

We might want to be picky about whether true/false values are in a JS or JSON context. So JS false would be `false`, but JSON false would be `"false"`. What do you think?

> The `"all_frames"` field specifies if the content script should be injected into all frames matching the specified URL requirements. If set to `false` it will only inject into the topmost frame. It can be used along with `"match_about_blank"` to inject into an `about:blank` frame. 

Author: ```suggestion
The `"all_frames"` field specifies if the content script should be injected into all frames matching the specified URL requirements. If set to `"false"` it will only inject into the topmost frame. It can be used along with `"match_about_blank"` to inject into an `about:blank` frame. 
```

To inject into other frames like `data:`, `blob:`, and `filesystem:`, set the `"match_origin_as_fallback"` to `true`. For details, see [Injecting in related frames][cs-inject-frames]

> `"all_frames"` Boolean

Reviewer: ```suggestion
`"all_frames"` - Boolean
```

> : _Optional_. Defaults to `false`, meaning that only the top frame is matched. If set to true, it will inject into all frames, even if the frame is not the topmost frame in the tab. Each frame is checked independently for URL requirements, it won't inject into child frames if the URL requirements are not met.

Reviewer: ```suggestion
: _Optional_. Defaults to `false`, meaning that only the top frame is matched. If set to true, it will inject into all frames, even if the frame is not the topmost frame in the tab. Each frame is checked independently for URL requirements. It won't inject into child frames if the URL requirements are not met.
```

> : _Optional_. Defaults to `false`, meaning that only the top frame is matched.If specified true, it will inject into all frames, even if the frame is not the topmost frame in the tab. Each frame is checked independently for URL requirements, it won't inject into child frames if the URL requirements are not met.

Reviewer 2: ```suggestion
: _Optional_. Defaults to `false`, meaning that only the top frame is matched. If set to true, it will inject into all frames, even if the frame is not the topmost frame in the tab. Each frame is checked independently for URL requirements, it won't inject into child frames if the URL requirements are not met.
```

`"match_about_blank"`- Boolean

> : _Optional_. Defaults to `false`. Whether the script should inject into an `about:blank` frame where the parent URL matches one of the patterns declared in `"matches"`.

Reviewer: ```suggestion
: _Optional_. Defaults to `false`. Specifies whether the script should inject into an `about:blank` frame when the parent URL matches one of the patterns declared in `"matches"`.
```

`"match_origin_as_fallback"` - Boolean

> : _Optional_. Defaults to `false`. Whether the script should inject in frames that were created by a matching origin, but whose URL or origin may not directly match the pattern. These include frames with different schemes, such as `about:`, `data:`, `blob:`, and `filesystem:`.

Reviewer: ```suggestion
: _Optional_. Defaults to `false`. Specifies whether the script should inject in frames that were created by a matching origin, but whose URL or origin may not directly match the pattern. These include frames with different schemes, such as `about:`, `data:`, `blob:`, and `filesystem:`.
```

## Run time and execution environment {: #world-timings }

> By default, content scripts are injected when the document and all resources are finished loading (`"document_idle"`), and live in a private isolated execution environment that isn't accessible to the page or other extensions.

Reviewer: What's `"document_idle"`?

Author: I removed this

> By default, content scripts are injected when the DOM is complete (`"document_idle"`), meaning all its resources have finished loading, and live in a private isolated execution environment that isn't accessible to the page or other extensions.

Reviewer 2: I could be wrong but (despite it being in the Chromium source) I'm not sure if "DOM is complete" is a good way of describing this accurately. It makes me think of [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event) which explicitly doesn't wait for images to load.

> [`"run_at"`][scripting-runat] - `document_start` | `document_end` | `document_idle`

Reviewer: Shouldn't the possible values be in quotes?

Reviewer: Since you define these below, I don't think you need them here. If you want them scannable, make the items below into bullets.

> : _Optional_. Specifies when the script should be injected into the page. It corresponds with the loading stated of [Document.readyState][mdn-ready-state]. Either `"document_start"` (the DOM is still loading), `"document_end"` (other resources are still loading), or `"document_idle"`, when the DOM and resources have finished loading.

Reviewer: ```suggestion
: _Optional_. Specifies when the script should be injected into the page. It corresponds with the loading stated of [Document.readyState][mdn-ready-state]. Must be one of `"document_start"` (the DOM is still loading), `"document_end"` (other resources are still loading), or `"document_idle"`, when the DOM and resources have finished loading.
```

Reviewer: >the loading stated of Document.readyState

What does this mean?

> : _Optional_. Specifies when the script should be injected into the page. It corresponds with the loading stated of [Document.readyState][mdn-ready-state]. Either `"document_start"` (the DOM is still loading), `"document_end"` (other resources are still loading), or `"document_idle"`, which corresponds to complete (the DOM and page resources have finished loading).

Reviewer 2: Same comment on use of "complete".

[`"world"`][scripting-world] - `ISOLATED` | `MAIN`

> : _Optional_. The JavaScript world for a script to execute within. Defaults to `"ISOLATED"`, which is the execution environment unique to the content script. Choosing `"MAIN"` world, means the script will share the execution environment with the host page's JavaScript.

Reviewer 2: Can we link to https://developer.chrome.com/docs/extensions/mv3/content_scripts/#isolated_world?

Author: I already had it in the aside below. It looked a bit redundant, but I'll add it to see if it fits

> : _Optional_. The JavaScript world for a script to execute within. Defaults to `"ISOLATED"`, which is the execution environment unique to the content script. Choosing `"MAIN"` world, means the script will share the execution environment with the host page's JavaScript. See [Work in isolated worlds][cs-worlds] to learn more.

Reviewer: ```suggestion
: _Optional_. The JavaScript world for a script to execute within. Defaults to `"ISOLATED"`, which is the execution environment unique to the content script. Choosing the `"MAIN"` world means the script will share the execution environment with the host page's JavaScript. See [Work in isolated worlds][cs-worlds] to learn more.
```

{% Aside 'warning' %}

> There are risks involved when using `"MAIN"` world. The host page can access and interfere with the injected script. See [Work in isolated worlds][cs-worlds] to learn more.

Reviewer: ```suggestion
There are risks involved when using the `"MAIN"` world. The host page can access and interfere with the injected script. See [Work in isolated worlds][cs-worlds] to learn more.
```
